### DISCUSSION TITLE
clangd checks are not works on infinite-loop and null dereference

### DISCUSSION THREAD

**Discussion by cacosub7** [clangd/clangd#1609] [Q&A] [answered]
When using clangd with clang-tidy enabled, I set .clang-tidy to enable all checks and the compiler provided the -Wall option. 
- .clang-tidy
```
Checks: '*'
WarningsAsErrors: '*'
```
- clangd version
clangd version 16.0.0 (https://github.com/llvm/llvm-project.git e52a41d0a2c1f880c0a99642ead09984448dd848)

However, it seems that clang-tidy still cannot detect the following two types of errors:

- infinite loop errors
clang-tidy checks `bugprone-infinite-loop`
```
bool CheckFoo(int foo)
{
    for (int i = 0; i < foo;)
    {
        if (foo % 3 == 0) continue;
        ++i;
    }
    //more complex code block
    return true; 
}
```
Obviously, if the value of foo meets the check then it will continue forever.
- Dereferencing of potentially null pointers
```
int TestNullPtr(int *p)
{
    printf("%d",*p);
    return 1;
}
```
Maybe the check of null pointer dereference is only available on `Clang Static Analyzer`?  BTW, i noticed  on the official website for Clang-Tidy that it supports clang static analyzer(`clang-analyzer-*`).
> clang-tidy has its own checks and can also run Clang Static Analyzer checks.

Regarding the above cases of infinite loops, has support been added? And for the second case of potentially null pointer dereference, which rule is it based on?

**Comment by HighCommander4** [comment]
Indeed, checks based on the static analyzer are not currently supported in clangd. See #905 for additional discussion and an outline of what would need to be done to support this.

**Reply by cacosub7** [reply]
Thank you for your reply. I have noticed that infinite loop detection in clang-tidy works in clangd. Specifically, the example I provided was not included in the list of examples shown in clang-tidy. Could you please provide more information on this point? Is the partial support for infinite loop detection related to the disabling of static analysis, or is it because the example of infinite loop that I provided is not covered in the logic of the detection algorithm? In the latter case, I would be willing to add the necessary detection logic myself to make this feature work. Please let me know if this is feasible and how I should proceed. Thank you in advance for your help

**Reply by HighCommander4** [reply]
It looks like the `bugprone-infinite-loop` check does not in fact use the static analyzer. So, clangd's behaviour with it should match clang-tidy's behaviour.

Are you seeing clangd's behaviour being different from clang-tidy's behaviour on a particular example?

**Reply by HighCommander4** [reply, marked as answer]
If you're asking about this loop from your original post in particular:

> ```
> bool CheckFoo(int foo)
> {
>     for (int i = 0; i < foo;)
>     {
>         if (foo % 3 == 0) continue;
>         ++i;
>     }
>     //more complex code block
>     return true; 
> }
> ```
> 
> Obviously, if the value of foo meets the check then it will continue forever.

Neither `clangd` nor `clang-tidy` issue the warning for me on this loop, i.e. they have the same behaviour.

If you're interested in extending the checker to try to catch more cases, please file an issue at https://github.com/llvm/llvm-project/issues and the developers of the checker can provide further guidance there.

**Reply by cacosub7** [reply]
Thank you again for your detailed response. I have tested the infinite loop example locally and both sides behave consistently. It seems that the current algorithm does not cover this case. I will further examine the algorithm logic to see if we can add a dead loop detection for the above case.
